FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+  // If SlowPath is true, then TopOffset is always zero.
+  int8_t TopOffset = 0;
+};
+
+
+class X87StackOptimization final : public FEXCore::IR::Pass {
+public:
+  X87StackOptimization() {
+    FEX_CONFIG_OPT(ReducedPrecision, X87REDUCEDPRECISION);
+    ReducedPrecisionMode = ReducedPrecision;
+  }
+  void Run(IREmitter* Emit) override;
+
+private:
+  bool ReducedPrecisionMode;
+  // FIXME(pmatos): copy from OpcodeDispatcher.h
+  [[nodiscard]]
+  uint32_t MMBaseOffset() {
+    return static_cast<uint32_t>(offsetof(Core::CPUState, mm[0][0]));
+  }
+  std::tuple<Ref, Ref> SplitF64SigExp(Ref Node);
+
+  // Top Management Helpers
+  /// Set the valid tag for Value as valid (if Valid is true), or invalid (if Valid is false).
+  void SetX87ValidTag(Ref Value, bool Valid);
+  // Generates slow code to load/store a value from the top of the stack
+  Ref LoadStackValueAtTop_Slow();
+  void StoreStackValueAtTop_Slow(Ref Value, bool SetValid = true);
+  // Generates slow code to load/store a value from an offset from the top of the stack
+  Ref LoadStackValueAtOffset_Slow(uint8_t Offset);
+  void StoreStackValueAtOffset_Slow(uint8_t Offset, Ref Value, bool SetValid = true);
+  // Update Top value in slow path for a pop
+  void UpdateTop4Pop_Slow();
+  void UpdateTop4Push_Slow();
+  // Synchronizes the current simulated stack with the actual values.
+  // Returns a new value for Top, that's synchronized between the simulated stack
+  // and the actual FPU stack.
+  Ref SynchronizeStackValues();
+  // Moves us from the fast to the slow path if ShouldMigrate is true.
+  void MigrateToSlowPathIf(bool ShouldMigrate);
+  // Top Cache Management
+  Ref GetTopWithCache_Slow();
+  Ref GetOffsetTopWithCache_Slow(uint8_t Offset);
+  void SetTopWithCache_Slow(Ref Value);
+  Ref GetX87ValidTag_Slow(uint8_t Offset);
+  // Resets fields to initial values
+  void Reset(bool AlsoSlowPath = true);
+
+  struct StackMemberInfo {
+    IR::OpSize SourceDataSize; // Size of SourceDataNode
+    IR::OpSize StackDataSize;  // Size of the data in the stack
+    IR::Ref SourceDataNode;    // Reference to the source location of the data.
+                               // In the case of a load, this is the source node of the load.
+                               // If it's not a load, then it's nullptr.
+    IR::Ref StackDataNode;     // Reference to the data in the Stack.
+    bool InterpretAsFloat {};  // True if this is a floating point value, false if integer
+  };
+
+  // StackData, TopCache need to be always properly set to ensure
+  // it reflects the current state of the FPU. This sync only makes sense while
+  // taking the fast path. Once in the slow path, these don't make sense anymore
+  // and we are syncing everything.
+
+  // Index on vector is offset to top value at start of block
+  // If slow path is true, then StackData is always empty.
+  FixedSizeStack<StackMemberInfo, 8> StackData;
+  using StackSlot = FixedSizeStack<StackMemberInfo, 8>::Slot;
```

Comment:
Could we just define the `enum class StackSlot` at top-level instead of nested in FixedSizeStack? That avoids special renaming.